File: common/src/main/java/dev/throwouterror/game/common/Entity.java
/*
 * Copyright (c) Creepinson
 */

package dev.throwouterror.game.common;

import org.joml.Matrix4f;

import java.io.Serializable;
import java.util.UUID;

/**
 * @author Theo Paris
 */
public class Entity implements Serializable {
    private Transform transform;
    private String type;
    private UUID id;

    public Entity(UUID id, String type, Transform transform) {
        this.id = id;
        this.type = type;
        this.transform = transform;
    }

    public Entity(Entity e) {
        this.transform = e.transform.clone();
    }

    public Matrix4f getScaleMatrix() {
        return new Matrix4f().scale(transform.getScale());
    }

    public Matrix4f getTranslationMatrix() {
        return new Matrix4f().translate(transform.getPosition());
    }

    public Matrix4f getRotation() {
        return transform.getRotation();
    }

    public Transform getTransform() {
        return this.transform;
    }

    public String getType() {
        return this.type;
    }

    public String getID() {
        return this.id.toString() + "-" + this.type;
    }


}

File: common/src/main/java/dev/throwouterror/game/common/Entity.kt
/*
 * Copyright (c) Creepinson
 */

package dev.throwouterror.game.common

import dev.throwouterror.game.common.networking.TextPacket
import dev.throwouterror.util.data.JsonUtils
import java.io.Serializable
import java.util.*

/**
 * @author Theo Paris
 */
open class Entity : Serializable {
    var transform: Transform = Transform()
        set
    val type: String
    var id: UUID

    constructor(id: UUID, type: String, transform: Transform) {
        this.id = id
        this.type = type
        this.transform = transform
    }

    constructor(e: Entity) {
        this.id = e.id
        this.type = e.type
        this.transform = e.transform.clone()
    }

    fun getID(): String {
        return this.id.toString() + "-" + this.type
    }

    override fun toString(): String {
        return getID()
    }

    fun toJson(): String {
        return JsonUtils.get().toJson(this)
    }

    fun toPacket(name: String): TextPacket {
        return TextPacket(name, toJson())
    }
}
